### PR TITLE
Add RefCounted class

### PR DESCRIPTION
**Pull request by temcguir**
This is used for shared state between threads.

 It will allow us to track and release Surface usage
 between a GL thread and the main thread.

Example of how it is being used [here](https://github.com/google/jetpack-camera-app/pull/132/files#diff-2c5c256dae5a6352542cb2bad6539b3bb4a36935bc7d845c014315339c4bd9ab)

### CODE REVIEW

File: core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+class RefCounted<T : Any>(
+    private val debugRefCounts: Boolean = false,
+    private val onRelease: (T) -> Unit = {}
+) {
+    private val refCounted = atomic(uninitialized<T>())
+
+    /**
+     * Initializes the ref-count managed object with the object being managed.
+     *
+     * This also initializes the implicit ref-count to 1.
```
Comment (avirakesh) [inline core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt:39]: nit: explicitly mention that `release` must be called after `initialize`

File: core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+     * Retrieves the underlying managed object, increasing the ref-count by 1.
+     *
+     * This increases the ref-count if the object has not already been released. If the object has
+     * been released, `null` is returned.
+     *
+     * All calls to this function must be paired with [release], unless `null` is returned.
+     */
+    fun acquire(): T? {
+        check(refCounted.value != uninitialized<T>()) {
+            "Ref-count managed object has not yet been initialized. Unable to acquire."
```
Comment (avirakesh) [inline core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt:68]: nit: "Unable to release."

File: core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+ * A thread-safe, lock-free class for managing the lifecycle of an object using ref-counting.
+ *
+ * The object that is ref-counted can be initialized late using [initialize].
+ *
+ * @param[debugRefCounts] whether to print debug log statements.
+ * @param[onRelease] a block that will be invoked once when the ref-count reaches 0.
+ */
+class RefCounted<T : Any>(
+    private val debugRefCounts: Boolean = false,
+    private val onRelease: (T) -> Unit = {}
```
Comment (avirakesh) [inline core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt:32]: From an API perspective, I would force this to be a required parameter. After all, Kotlin doesn't really need reference counting if onRelease is a no-op. People coming from other languages may try to use RefCouted where `shared_ptr` or `sp` would typically be used (even though it is not needed).

**Review by avirakesh — APPROVED** [review]
Generally LGTM! I really wish Kotlin had better resource support handling, but this seems like a decent way to go about it!

File: core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt (hunk 2 above)
Comment (temcguir) [reply core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt:71]: Good catch. Made this change in the `release()` function.

File: core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt (hunk 1 above)
Comment (temcguir) [reply core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt:39]: Done.

File: core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt (hunk 3 above)
Comment (temcguir) [reply core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt]: Yeah that's probably better- changed it to be required.

The only case I can think of where you would want a no-op `onRelease` is for flow control. That is, wrap the object and perform tasks only if you can `acquire()`. However, there are probably better ways to do that, and you can always pass in a no-op `onRelease` if that's what you want.

**Comment by temcguir** [comment]
Thanks for the review, Avi!

**Pull request merged**